### PR TITLE
parse define

### PR DESCRIPTION
**Pull request by jitwit**
Adds parsing for functions defined like `(define f (lambda (x ...) b))`. 

I wanted to add non-function definitions, too, but wasn't successful. I'm guessing they should be added to the export environment with `add-top-level`.  I got things like `(define a 0)` to work but not `(define b (cons a a))`. How might I get them to work?

### CODE REVIEW

File: schism/compiler.ss [hunk 1]
```diff
@@ -392,6 +394,34 @@
...
+         (eq? (car def) 'define)
+         (pair? (cdr def))
+         (or (pair? (cadr def))
+             (and (pair? (cddr def))
+                  (pair? (caddr def))
+                  (eq? 'lambda (caaddr def))))))
+  (define (function-formals def)
+    (if (pair? (cadr def))
+        (cdadr def)
+        (car (cdaddr def))))
```
Comment (eholk) [inline schism/compiler.ss:419]: For each of these functions, in the `lambda` cause, could you add a check to make sure there is a `'lambda` in the right place? That will help prevent accidental matches as we expand support for more complex expressions.

Comment (jitwit) [reply schism/compiler.ss]: Makes sense! hopefully the added checks look reasonable?

**Comment by jitwit** [comment]
I think I understand the gist of your approach. Once I realized the simplest approach didn't work for complex expressions, I thought maybe something along the lines of `letrec*` or another pass with a some environment would be needed. I'll take a look at things that use `global-get` and `global-set` to try and understand how to use them and would for sure like to try and get it working

File: schism/compiler.ss (hunk 1 above)
Comment (eholk) [reply schism/compiler.ss]: Looks good!

**Comment by eholk** [comment]
Oh, good point, there might be some trickiness around making sure each initializer runs in the right environment, to make sure they don't refer to things that aren't defined yet. 

I forgot that Schism already supports the `global.get` and `global.set` instructions. It looks like we use them for interning constants. That's related to top-level defines, so looking at the way we do interning already might be helpful.

**Comment by jitwit** [comment]
I thought a bit more about this today. Before doing constants I realized that `define`s in inner bodies won't work either. I think what can be done there is:
* split the body by (0+) definitions and (1+) expressions and add lexicals for the defined names before running `parse-body` 
* add a clause in `parse-expr` for definitions to parse them

What I've hit is in `simplify-expr`, an error for unrecognized expr. It seems like the next step is adding a tag and a way to simplify the definitions? Is this the right approach?

**Comment by eholk** [comment]
Thanks for the other PR with support for define. I think the top tag works
okay.

**Pull request merged**